zmq.hpp

```diff
@@ -615,8 +620,340 @@ inline void swap(context_t &a, context_t &b) ZMQ_NOTHROW {
     a.swap(b);
 }
 
+#ifdef ZMQ_CPP11
+struct send_result
+{
+    size_t size;    // message size in bytes
+    bool success;
+};
+
+struct recv_result
+{
+    size_t size;    // message size in bytes
+    bool success;
+};
+
+struct recv_buffer_result
+{
+    size_t size;    // number of bytes written to buffer
+    size_t untruncated_size;  // untruncated message size in bytes
+    bool success;
+
+    ZMQ_NODISCARD bool truncated() const noexcept
+    {
+        return size != untruncated_size;
+    }
+};
+
+enum class send_flags : int
+{
+    none = 0,
+    dontwait = ZMQ_DONTWAIT,
+    sndmore = ZMQ_SNDMORE
+};
+
+constexpr send_flags operator|(send_flags a, send_flags b) noexcept
+{
+    return static_cast<send_flags>(static_cast<int>(a) | static_cast<int>(b));
+}
+
+enum class recv_flags : int
+{
+    none = 0,
+    dontwait = ZMQ_DONTWAIT
+};
+
+constexpr recv_flags operator|(recv_flags a, recv_flags b) noexcept
+{
+    return static_cast<recv_flags>(static_cast<int>(a) | static_cast<int>(b));
+}
+
+// mutable_buffer, const_buffer and buffer are based on
+// the Networking TS specification, draft:
+// http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/n4771.pdf
+
+class mutable_buffer
+{
+  public:
+    mutable_buffer() noexcept : _data(nullptr), _size(0) {}
+    mutable_buffer(void *p, size_t n) noexcept : _data(p), _size(n)
+    {
+        assert(p != nullptr || n == 0);
+    }
+
+    void *data() const noexcept { return _data; }
+    size_t size() const noexcept { return _size; }
+    mutable_buffer &operator+=(size_t n) noexcept
+    {
+        // (std::min) is a workaround for when a min macro is defined
+        const auto shift = (std::min)(n, _size);
+        _data = static_cast<char *>(_data) + shift;
+        _size -= shift;
+        return *this;
+    }
+
+  private:
+    void *_data;
+    size_t _size;
+};
+
+inline mutable_buffer operator+(const mutable_buffer &mb, size_t n) noexcept
+{
+    return mutable_buffer(static_cast<char *>(mb.data()) + (std::min)(n, mb.size()),
+                          mb.size() - (std::min)(n, mb.size()));
+}
+inline mutable_buffer operator+(size_t n, const mutable_buffer &mb) noexcept
+{
+    return mb + n;
+}
+
+class const_buffer
+{
+  public:
+    const_buffer() noexcept : _data(nullptr), _size(0) {}
```

Comment:
The constructors and data and size member functions could be constexpr, as well as the factory functions below. (Not sure about the operators)

Reply:
That's true. I will see how constexpr I can make it.

Reply:
Only if you have time, this could be added later on as well. I just noticed this since you used constexpr at some other places.

Sorry I didn't make clear in my review comments what were just ideas, and what I feel should be changed right now.